solutions/usaco-1159.mdx

```diff
@@ -0,0 +1,142 @@
+---
+id: usaco-1159
+source: USACO Silver 2021 December
+title: Connecting Two Barns
+author: Nathan Gong
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_prob2_gold_open21.html)
+
+## Implementation - DFS + Two Pointers
+
+**Time Complexity:** $\mathcal{O}(T\cdot (N + M))$
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class ConnectingTwoBarns {
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		int t = io.nextInt();
+		for (int test = 0; test < t; test++) {
+			int n = io.nextInt();
+			int m = io.nextInt();
+
+			List<List<Integer>> adj = new ArrayList<>();
+			for (int i = 0; i < n; i++) {
+		   		adj.add(new ArrayList<>());
+			}
+			for (int i = 0; i < m; i++) {
+				int a = io.nextInt() - 1;
+				int b = io.nextInt() - 1;
+				adj.get(a).add(b);
+				adj.get(b).add(a);
+			}
+
+			// stores the component each node is in
+			int[] visited = new int[n];
+			// stores a list of nodes for each component
+			List<List<Integer>> components = new ArrayList<>();
+
+			int numComponents = 0;
+			Arrays.fill(visited, -1);
+			for (int i = 0; i < n; i++) {
+				// run DFS on node if it hasn't been visited
+				if (visited[i] == -1) {
+					components.add(new ArrayList<>());  // add new component
+					Stack<Integer> stack = new Stack<>();
+					stack.push(i);
+					while (!stack.isEmpty()) {
+						int curr = stack.pop();
+						if (visited[curr] != -1) continue;
+						// add node to component
+						components.get(numComponents).add(curr);
+						// mark node as visited
+						visited[curr] = numComponents;
+
+						for (int neighbor : adj.get(curr)) {
+							stack.push(neighbor);
+						}
+					}
+					numComponents++;
+				}
+			}
+			// sort nodes in each component by position
+			for (List<Integer> component : components) {
+				Collections.sort(component);
+			}
```

Comment:
Wait if you do this, the solution would be `O(m + n log n)`. I thought the editorial says it's `O(n + m)`, since the order of the DFS guarantees that the nodes in each component are in sorted order 😛.